util/topsql/tracecpu/profile.go

```diff
@@ -0,0 +1,421 @@
+	}
+	pprof.SetGoroutineLabels(ctx)
+	return ctx
+}
+
+// SetSQLAndPlanLabels sets the SQL and plan digest label into the goroutine.
+func SetSQLAndPlanLabels(ctx context.Context, sqlDigest, planDigest string) context.Context {
+	ctx = pprof.WithLabels(ctx, pprof.Labels(labelSQLDigest, sqlDigest, labelPlanDigest, planDigest))
+	pprof.SetGoroutineLabels(ctx)
+	return ctx
+}
+
+func (sp *topSQLCPUProfiler) startExportCPUProfile(w io.Writer) error {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	if sp.mu.ept != nil {
+		return errors.New("cpu profiling already in use")
+	}
+	sp.mu.ept = &exportProfileTask{w: w}
+	return nil
+}
+
+func (sp *topSQLCPUProfiler) stopExportCPUProfile() error {
+	sp.mu.Lock()
+	ept := sp.mu.ept
+	sp.mu.ept = nil
+	sp.mu.Unlock()
+	if ept.err != nil {
+		return ept.err
+	}
+	if w := ept.w; w != nil && ept.cpuProfile != nil {
+		sp.removeLabel(ept.cpuProfile)
+		return ept.cpuProfile.Write(w)
+	}
+	return nil
+}
+
+// removeLabel uses to remove labels for export cpu profile data.
+// Since the sql_digest and plan_digest label is strange for other users.
+// If `variable.EnablePProfSQLCPU` is true means wanto keep the `sql` label, otherwise, remove the `sql` label too.
+func (sp *topSQLCPUProfiler) removeLabel(p *profile.Profile) {
+	if p == nil {
+		return
+	}
+	keepLabelSQL := variable.EnablePProfSQLCPU.Load()
+	for _, s := range p.Sample {
+		for k := range s.Label {
+			switch k {
+			case labelSQL:
+				if !keepLabelSQL {
+					delete(s.Label, k)
+				}
+			case labelSQLDigest, labelPlanDigest:
+				delete(s.Label, k)
+			}
+		}
+	}
+}
+
+type exportProfileTask struct {
+	cpuProfile *profile.Profile
+	err        error
+	w          io.Writer
+}
+
+func (t *exportProfileTask) mergeProfile(p *profile.Profile) {
+	if t.err != nil || p == nil {
+		return
+	}
+	ps := make([]*profile.Profile, 0, 2)
+	if t.cpuProfile != nil {
+		ps = append(ps, t.cpuProfile)
+	}
+	ps = append(ps, p)
+	t.cpuProfile, t.err = profile.Merge(ps)
+}
+
+// ProfileHTTPHandler is same as pprof.Profile.
+// The difference is ProfileHTTPHandler uses tracecpu.StartCPUProfile/StopCPUProfile to fetch profile data.
+func ProfileHTTPHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	sec, err := strconv.ParseInt(r.FormValue("seconds"), 10, 64)
+	if sec <= 0 || err != nil {
+		sec = 30
+	}
+
+	if durationExceedsWriteTimeout(r, float64(sec)) {
+		serveError(w, http.StatusBadRequest, "profile duration exceeds server's WriteTimeout")
+		return
+	}
+
+	// Set Content Type assuming StartCPUProfile will work,
+	// because if it does it starts writing.
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	err = StartCPUProfile(w)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+		return
+	}
+	// TODO: fix me.
```

Comment:
This can be fixed by always starts a 1 second profiling one by one, but to aggregate (merge) multiple profiles into one according to the precision.

Reply:
yes, already update the comment.